src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -249,53 +274,58 @@ static P11Key derivePBEKey(Token token, PBEKeySpec keySpec, String algo)
                     //
                     int inputLength = (password == null) ? 0 : password.length;
-                    char[] expPassword = new char[inputLength * 2 + 2];
+                    expPassword = new char[inputLength * 2 + 2];
                     for (int i = 0, j = 0; i < inputLength; i++, j += 2) {
                         expPassword[j] = (char) ((password[i] >>> 8) & 0xFF);
                         expPassword[j + 1] = (char) (password[i] & 0xFF);
                     }
-                    password = expPassword;
                 }
                 ckMech = new CK_MECHANISM(kdfData.kdfMech,
-                        new CK_PBE_PARAMS(password, salt, itCount));
+                        new CK_PBE_PARAMS(expPassword, salt, itCount));
             }
 
-            long keyType = getKeyType(kdfData.keyAlgo);
-            CK_ATTRIBUTE[] attrs = new CK_ATTRIBUTE[
-                    switch (kdfData.op) {
-                        case ENCRYPTION, AUTHENTICATION -> 4;
-                        case GENERIC -> 5;
-                    }];
-            attrs[0] = new CK_ATTRIBUTE(CKA_CLASS, CKO_SECRET_KEY);
-            attrs[1] = new CK_ATTRIBUTE(CKA_VALUE_LEN, keySize >> 3);
-            attrs[2] = new CK_ATTRIBUTE(CKA_KEY_TYPE, keyType);
-            switch (kdfData.op) {
-                case ENCRYPTION -> attrs[3] = CK_ATTRIBUTE.ENCRYPT_TRUE;
-                case AUTHENTICATION -> attrs[3] = CK_ATTRIBUTE.SIGN_TRUE;
-                case GENERIC -> {
-                    attrs[3] = CK_ATTRIBUTE.ENCRYPT_TRUE;
-                    attrs[4] = CK_ATTRIBUTE.SIGN_TRUE;
-                }
-            }
-            CK_ATTRIBUTE[] attr = token.getAttributes(
-                    O_GENERATE, CKO_SECRET_KEY, keyType, attrs);
+            CK_ATTRIBUTE[] attrs = new CK_ATTRIBUTE[] {
+                    new CK_ATTRIBUTE(CKA_CLASS, CKO_SECRET_KEY),
+                    new CK_ATTRIBUTE(CKA_VALUE_LEN, keySize >> 3),
+                    new CK_ATTRIBUTE(CKA_KEY_TYPE,
+                            getPKCS11KeyType(kdfData.keyAlgo)),
+                    kdfData.encrypt_or_sign_true,
+            };
+            CK_ATTRIBUTE[] attr = token.getAttributes(O_GENERATE,
+                    CKO_SECRET_KEY, getKeyType(kdfData.keyAlgo), attrs);
             long keyID = token.p11.C_GenerateKey(session.id(), ckMech, attr);
-            return (P11Key)P11Key.secretKey(
-                    session, keyID, kdfData.keyAlgo, keySize, attr);
+            return (P11Key) P11Key.pbeKey(session, keyID, algo,
+                    keySize, attr, password, salt, itCount);
         } catch (PKCS11Exception e) {
             throw new InvalidKeySpecException("Could not create key", e);
         } finally {
             token.releaseSession(session);
         }
     }
 
+    private static PBEKeySpec getPbeKeySpec(PBEKey pbeKey) {
+        int keyLength = 0;
+        if ("RAW".equals(pbeKey.getFormat())) {
```

Comment:
Have we checked that all these PBE-derived secret keys that we take out of the NSS Software Token do not have any wrapping/formatting? Do we get any guarantees from the PKCS#11 standard to extrapolate this to non-NSS tokens? What would be the consequences of not setting a keyLength in this method? (i.e. can't we get that information from the algorithm and the KDFMap anyways in a more reliable way?)

Reply:
I've checked that ::getEncoded for Secret Keys in the NSS Software Token always returns their CKA_VALUE attribute, which is a raw key (without any formatting) for all key types. This is backed up by the PKCS#11 standard. The consequences of not setting a keyLength might be that we cannot derive a PKBDF2 key out of an already derived PBKDF2 key (i.e. one PBKDF2 key from SunJCE that is being translated for use in SunPKCS11) because the key length map is not in the KDFMap for these algorithms.